VERSIONING.md

```diff
@@ -0,0 +1,157 @@
+# Versioning
+
+This document covers Vector's versioning and what it means as a user of Vector.
+
+**Please note, Vector is currently in its pre-1.0 phase and quickly approaching
+1.0. Minor version increments can introduce breaking changes during this phase.
+Please see the [FAQ](#faq) section for more info.**
+
+<!-- MarkdownTOC autolink="true" style="ordered" indent="   " -->
+
+1. [Convention](#convention)
+1. [Public API](#public-api)
+   1. [Areas that *are* covered](#areas-that-are-covered)
+      1. [Intended for *public* consumption](#intended-for-public-consumption)
+      1. [Intended for *private* consumption](#intended-for-private-consumption)
+   1. [Areas that are *NOT* covered](#areas-that-are-not-covered)
+1. [FAQ](#faq)
+   1. [How often is Vector released?](#how-often-is-vector-released)
+   1. [How does Vector treat patch and minor versions?](#how-does-vector-treat-patch-and-minor-versions)
+   1. [How does Vector treat major versions \(breaking changes\)?](#how-does-vector-treat-major-versions-breaking-changes)
+   1. [How does Vector treat pre-1.0 versions?](#how-does-vector-treat-pre-10-versions)
+   1. [Release cadence](#release-cadence)
+
+<!-- /MarkdownTOC -->
+
+## Convention
+
+Vector adheres to the [Semantic Versioning 2.0] convention. In summary:
+
+* Versions follow the `MAJOR.MINOR.PATCH` format (i.e., `2.5.1`)
+* `PATCH` increments only when backward compatible bug fixes are introduced.
+* `MINOR` increments only when new, backward compatible functionality is introduced.
+* `MAJOR` increments if any backwards incompatible changes are introduced.
+* Pre `1.0` (major version `0`) is for initial development and `MINOR` version bumps can introduce breaking changes.
+
+## Public API
+
+Semantic Versioning hinges on Vector's defintion of "public API". By nature of
+Vector - a tool that collects, processes, and routes data from disparate systems
+- it has a very large public surface area. It's not immediately obvious which
+parts are covered under our versioning contract and how they're covered. This
+section aims to remove all ambiguity in this area.
+
+### Areas that *are* covered
+
+The following Vector areas are covered in Vector's definition of public API.
+
+#### Intended for *public* consumption
+
+The follow Vector areas are inteded for *public* consumption (consumption by
+anything other than Vector itself). Backward incompatble changes will trigger
+a major version increment.
+
+* [CLI]
+  * The root [`vector` command], its flags and exit code
+  * The [`vector validate` subcommand], the flags exit codes
```

Comment:
I think we can simplify this to just:

```
* [CLI]
  * Flags
  * Exit codes
```

And remove the specific notes about `vector` and `vector validate` since those guarantees would apply to all commands.

Reply:
Sure, but these are the only commands we want establish a public contract with. We should be able to change the others as necessary (rename, etc).

Reply:
Ah, I see. My opinion is that all subcommands should be covered by the policy. I think it'd be confusing to rename one without a major release.

Reply:
Is it worth a major version bump to rename the `list` command? That's for human interaction and we shouldn't be breaking contracts by doing that. Hence why I don't think it should be covered.

Reply:
I think it's an expectation thing. If I upgrade from 0.15.0 to 0.15.1 I would very much not expect a subcommand to be renamed; regardless of whether it is meant for humans or machines.